### ISSUE TITLE
Mapping suggestions

### ISSUE THREAD

**Issue by ottopiramuthu** [Jelmerro/Vieb#185]
**Checklist**
- [x] I searched the `:help` documentation using `/` for helpful information
- [x] I have read the [frequently asked questions](https://github.com/Jelmerro/Vieb/blob/master/FAQ.md)
- [x] I did not find any similar issues in the [issue tracker](https://github.com/Jelmerro/Vieb/issues)

**Addition or change**
`showcmd` is nice, but it doesn't show the user what keys they can press _next_.
The plugin Surfingkeys gives suggestions for mappings (when they are more than one character long) in the bottom right corner.
For example, when I type `z`, this shows up:

![image](https://user-images.githubusercontent.com/49997309/117556495-984ff980-b037-11eb-8ce7-b63587e8fd2f.png)

The key(s) I have pressed so far are highlighted, and next to each mapping is a short description.
These descriptions are defined/can be changed by the user in the settings:

![image](https://user-images.githubusercontent.com/49997309/117556536-05638f00-b038-11eb-937b-b78de1cf0173.png)

It would be nice to have an option that enables a similar suggestion box for mappings in Vieb.
I'm not sure how the short descriptions should be handled, so until that is figured out I'm fine with the descriptions just being the set of actions bound to each mapping. For example, if I press `g` in normal mode with this enabled, one of the lines in the suggestion box would be `gg <action.scrollTop>`, where the first `g` is in a different color than the second.

**Alternatives considered**
You can look at all mappings at once with `:map!`, but having a quick list of the mappings you can run now is very handy.

**Comment by Jelmerro** [comment]
Let me know if [this](https://github.com/Jelmerro/Vieb/commit/42aaa16b51f60eda52e97e8c6f41e3b80df2ddb9) is what you had in mind. The only thing I'm not sure about is if we should enable this by default, and if so, how many suggestions to show. It's currently enabled with 50 entries, but maybe we should just cap it at a low number like 5 by default.

**Comment by ottopiramuthu** [comment]
This works fantastically on Windows. I will test it on Linux when I get the chance. Thanks to this the Qutebrowser [example config ](https://github.com/Jelmerro/Vieb/issues/83#issuecomment-852543382) will feel less empty :)

 I prefer not limiting suggestions that much since it makes mappings less discoverable and I don't mind the large suggestions box. If you're willing to implement it, a scrollbar for if the suggestions go off screen would be nice. 

I think such a helpful feature should be on by default (I mean it's almost as useful as suggestions in Explore/Command mode). 

As a minor suggestion it would be nice to have a separate timeout setting for `mapsuggest` similar to `guihidetimeout`.

**Comment by Jelmerro** [comment]
A different timeoutlen would be kind of weird, because the box now just stays as long as it's relevant. If you want to use those mappings you need to keep the keys stored longer, otherwise you would have these suggestions, but you can't use them until you click the first key again because it expired even though the box is still shown. I do agree that the box isn't all that useful with the default timeoutlen, but that's why I included the the suggestion to increase or even disable it in the help for mapsuggest. Besides, it's not like timeoutlen does much else besides just caching the keys longer: the only other function is remembering which tabs you visited recently (as in viewed properly, instead of scrolled past).

I will look into making it scrollable, currently it just overflows if it gets too long, with isn't all that useful indeed. I will also keep or even increase the current limit, depending on how nice the scrolling works.

**Comment by Jelmerro** [comment]
It's now by default scrollable, in the process I also made some small fixes to notifications in general, which weren't always positioned correctly in different themes. I've increased the map suggest limit to the max value as the scrolling is really nice if you are using the mouse. I agree that these suggestions are nice to have by default. They are also easy to turn off if you don't want them at all with `:set mapsuggest=0`.

**Comment by ottopiramuthu** [comment]
I've been thinking about how the short descriptions could be implemented. Could those be put in an inline comment where the respective mappings are in the viebrc? For example,
```
nmap go <action.toExploreMode> " open here
```
would show
```
go        open here
```
in the map suggestions.

**Comment by Jelmerro** [comment]
That would only work for mappings from inside the viebrc? Besides, it would make it possible to hide the functionality of specific keys by intentionally giving it a description that doesn't match with the functionality. I'm not saying that would be common, but it's even possible to unintentionally do this to yourself, by placing a comment there about the reason for a specific mapping using those keys, instead of commenting about what it will do when executed. What I'm trying to say, is that without explaining this to users in depth, it's unlikely that they will know the function of commenting on the same line. Regardless, I think it's also weird that comments are interpreted when reading a config file, it should be the one thing that is completely up to the user to create however they want without worrying about it being interpreted.

The next release will make it possible to greatly reduce the strictness of action names, by allowing different casing and making the `action.` prefix optional (at least for actions, and for pointer `p.` is enough). This should help reducing the overhead of action names when mapping to them, which in turn should make it much more clear what is actually happening.

Finally, this is in a way already possible by defining a custom command to do whatever it is you want to do, and executing it using the `<:my-custom-command>` syntax. This does not require you to switch modes, and will reduce the amount of actions even further combined with the above. For example:
```vim
command new_tab_from_clipboard <action.openNewTab><action.openFromClipboard>
map pp <:new_tab_from_clipboard>
```
This will show as `pp <:new_tab_from_clipboard>` with the first p selected after pressing one p.
Starting from the next release, this could even be simplified to:
```vim
command new_tab_from_clipboard <openNewTab><openFromClipboard>
map pp <:new_tab_from_clipboard>
```
I think the above is simpler to explain, without the downsides of using comments for this. I hope that you understand why I don't like the comment suggestion. If you think we really need custom names for mapstrings, we could think of other ways to allow the labels to be customized.

**Comment by ottopiramuthu** [comment]
command new_tab_from_clipboard <openNewTab><openFromClipboard>
    map pp <:new_tab_from_clipboard>
This looks super clean. I don't need custom names for mapstrings are needed if you can define commands like this. I was thinking short descriptions could make the transition from different browsers/plugins easier, but custom commands would be enough for most cases.

Would you mind me making custom commands for the example viebrcs of browsers/extensions that have names for commands (Qutebrowser, Vimium, Tridactyl) so new users will see the commands they are familiar with or is that too much?

**Comment by Jelmerro** [comment]
Sure, go for it! Dashes between commands are also supported, if that's more in line with what qutebrowser and others are doing. Also feel free to use the new simpler action invoke syntax if that makes things easier. The latter was something I was already considering actually.

**Comment by ottopiramuthu** [comment]
There is a problem: qutebrowser has commands with spaces between them (its commands have optional flags and arguments). For example, `yank` is used for yanking many kinds of things. There are `yank title`, `yank url`, etc. How would you like to handle those cases? Underscores?

**Comment by Jelmerro** [comment]
I guess underscores are the best in that case, spaces are pretty much the only character that isn't allowed inside custom command names.

**Comment by ottopiramuthu** [comment]
I've tested it on Linux and it works well.

**Comment by ottopiramuthu** [comment]
I've been working on the updated qutebrowser viebrc with custom commands and [here](https://gist.github.com/ottopiramuthu/c5844929d8c84bee1c59cb7bc198e88c) is what I have so far. Everything looks fine to me, but I'm getting [these](https://gist.github.com/ottopiramuthu/20cda725f87c8306fd83acd784f4c0af) errors on startup and I'm not sure what could be happening. Do you know what could be wrong with the viebrc?

**Comment by Jelmerro** [comment]
This is only when running from source? I think I know what is happening, that's due to a recent change I made. I'll rework that in a few hours.

**Comment by ottopiramuthu** [comment]
Yes this happens when I source the file only.

**Comment by Jelmerro** [comment]
Let me know if this issue is now fixed with the latest code.

Also in your example config, I would recommend to use `command!` instead of `command` to allow the config to be reloaded without errors (the extra `!` allows you to overwrite commands).

**Comment by ottopiramuthu** [comment]
I was wondering whether using `command!` would be okay... I'll let you know if it works ASAP.

**Comment by ottopiramuthu** [comment]
With your latest commits this works. Let me know if there are any errors/inconsistencies with Qutebrowser that I missed.

**Comment by Jelmerro** [comment]
Implemented in the latest release 5.1.0 release.
